cpp/src/arrow/compute/kernels/isin.cc

```diff
@@ -0,0 +1,325 @@
+#include "arrow/util/string_view.h"
+#include "arrow/visitor_inline.h"
+
+namespace arrow {
+
+using internal::checked_cast;
+using internal::DictionaryTraits;
+using internal::HashTraits;
+
+namespace compute {
+
+class IsInKernelImpl : public UnaryKernel {
+  virtual Status Compute(FunctionContext* ctx, const Datum& left, Datum* out) = 0;
+
+ public:
+  // \brief Check if value in both arrays or not and returns boolean values/null
+  Status Call(FunctionContext* ctx, const Datum& left, Datum* out) override {
+    DCHECK_EQ(Datum::ARRAY, left.kind());
+    RETURN_NOT_OK(Compute(ctx, left, out));
+    return Status::OK();
+  }
+
+  std::shared_ptr<DataType> out_type() const override { return boolean(); }
+
+  virtual Status Reset() = 0;
+  virtual Status ConstructRight(FunctionContext* ctx, const Datum& right) = 0;
+};
+
+// ----------------------------------------------------------------------
+// Using a visitor create a memo_table_ for the right array
+// TODO: Implement for small lists
+
+template <typename T, typename Scalar>
+struct MemoTableRight {
+  Status VisitNull() { return Status::OK(); }
+
+  Status VisitValue(const Scalar& value) {
+    memo_table_->GetOrInsert(value);
+    return Status::OK();
+  }
+
+  Status Reset() {
+    memo_table_.reset(new MemoTable(0));
+    return Status::OK();
+  }
+
+  Status Append(FunctionContext* ctx, const Datum& right) {
+    const ArrayData& right_data = *right.array();
+    right_null_count = right_data.GetNullCount();
+    return ArrayDataVisitor<T>::Visit(right_data, this);
+  }
+
+  using MemoTable = typename HashTraits<T>::MemoTableType;
+  std::unique_ptr<MemoTable> memo_table_;
+  int64_t right_null_count;
+};
+
+// ----------------------------------------------------------------------
+
+template <typename Type, typename Scalar>
+class IsInKernel : public IsInKernelImpl {
+ public:
+  IsInKernel(const std::shared_ptr<DataType>& type, MemoryPool* pool)
+      : type_(type), pool_(pool) {}
+
+  // \brief if null count in right is not 0, then append true to the
+  // BooleanBuilder when left array has a null, else append null.
+  Status VisitNull() {
+    if (right_null_count != 0) {
+      bool_builder_.UnsafeAppend(true);
+    } else {
+      bool_builder_.UnsafeAppendNull();
+    }
+    return Status::OK();
+  }
+
+  // \brief Iterate over the left array using another visitor.
+  // In VisitValue, use the memo_table_ (for right array) and check if value
+  // in left array is in the memo_table_. Append the true to the
+  // BooleanBuilder if condition satisfied, else false.
+  Status VisitValue(const Scalar& value) {
+    bool_builder_.UnsafeAppend(memo_table_->Get(value) != -1);
+    return Status::OK();
+  }
+
+  Status Reset() override {
+    bool_builder_.Reset();
+    return Status::OK();
+  }
+
+  Status Compute(FunctionContext* ctx, const Datum& left, Datum* out) override {
+    const ArrayData& left_data = *left.array();
+    RETURN_NOT_OK(bool_builder_.Reserve(left_data.length));
```

Comment:
you end up allocating memory twice.  Once here and once using the PrimitiveAllocatingKernel.  I think you want to use FirstTimeBitmapWriter instead https://github.com/apache/arrow/blob/ceaed810e6e7a654d0f45db4c8e2f4bae7edefe5/cpp/src/arrow/util/bit-util.h#L562

Reply:
I have used FirstTimeBitmapWriter.

Reply:
@emkornfield  please let me know if any other change is needed.